### PR TITLE
[AT-40][Add] Preliminary Logging CorrelationID

### PR DESCRIPTION
**Pull request by martinstibbe**

### CODE REVIEW

File: edgegrid/log.go [hunk 1]
```diff
@@ -12,41 +12,45 @@
...
+
 	log "github.com/sirupsen/logrus"
 )
 
 var logBuffer *bufio.Writer
 var LogFile *os.File
 var EdgegridLog *log.Logger
 
+// LogCorrelationID ID for header and footer of log file outputs
+var LogCorrelationID *string
```
Comment (zstlaw) [inline edgegrid/log.go:33]: Shared variable will have trouble with concurrent calls.

Comment (ssrirama-aka) [reply edgegrid/log.go:33]: The same comments made in the provider AT-40 pull request apply here: I think the library should also be following a similar  [level][api][guid] <message> format. I would expect the common logging functions to be placed in the library itself and utilized in the provider.

File: edgegrid/log.go [hunk 2]
```diff
@@ -85,24 +98,53 @@ func LogMultilinef(f func(formatter string, args ...interface{}), formatter stri
...
-        b, err := httputil.DumpRequestOut(req, body)
-        if err == nil {
-                LogMultiline(EdgegridLog.Traceln, string(b))
-        }
+	if req == nil {
+		return
+	}
+	b, err := httputil.DumpRequestOut(req, body)
+	if err == nil {
+		LogMultiline(EdgegridLog.Traceln, string(b))
```
Comment (zstlaw) [inline edgegrid/log.go:106]: Isn't this logging the same request two different ways?  One as a single line out and another per line of the output?  I like logging the request as a single action but am unable to grok why we still need it the old way as well.  Can you explain?

**Comment by edglynes** [comment]
We don't need both. It won't hurt to always have a correlation ID.  There is additional work to be done if the logic is simplified.  All existing references will need to be tweaked including GTM and DNS. Since this is a library, we either need to require a correlation id, generate one if none provided or handle cases with and without.

**Comment by zstlaw** [comment]
Looks like some of my comments did not post from last week.  I didn't see any more glaring issues but I wasn't sure why some of the service calls pass empty log correlation ids.

File: edgegrid/log.go [hunk 3]
```diff
@@ -85,24 +98,53 @@ func LogMultilinef(f func(formatter string, args ...interface{}), formatter stri
...
+}
+
+func PrintHttpRequestCorrelation(req *http.Request, body bool, correlationid string) {
+
+	if req == nil {
+		return
+	}
+	b, err := httputil.DumpRequestOut(req, body)
+	if err == nil {
+		LogMultiline(EdgegridLog.Traceln, string(b))
```
[the remark this answers is not in the document]
Comment (martinstibbe) [reply edgegrid/log.go:118]: Ok I can remove these that was initial testing of wrapped logs I broke it out to separate function to allow incremental migration once correlation concept catches on

File: edgegrid/log.go (hunk 2 above)
Comment (martinstibbe) [reply edgegrid/log.go:106]: Ok I can remove these that was initial testing of wrapped logs I broke it out to separate function to allow incremental migration once correlation concept catches on

**Pull request merged**